src/dbnode/persist/fs/wide/entry_checksum_mismatch_checker.go

```diff
@@ -0,0 +1,284 @@
+		)
+	})
+	return err
+}
+
+func (c *entryChecksumMismatchChecker) ComputeMismatchForEntry(
+	indexEntry schema.IndexEntry,
+) ([]ReadMismatch, error) {
+	hash := c.decodeOpts.IndexEntryHasher()
+	checksum := hash.HashIndexEntry(indexEntry)
+	entry := entryWithChecksum{entry: indexEntry, idChecksum: checksum}
+	c.mismatches = c.mismatches[:0]
+	if c.exhausted {
+		// NB: no remaining batches in the index checksum block; any further
+		// elements are mismatches (missing from primary).
+		return c.entryMismatches(entry), nil
+	}
+
+	if !c.started {
+		c.started = true
+		if !c.blockReader.Next() {
+			// NB: no index checksum blocks available; any further
+			// elements are mismatches (missing from primary).
+			c.exhausted = true
+			return c.entryMismatches(entry), nil
+		}
+
+		c.batchIdx = 0
+	}
+
+	batch := c.blockReader.Current()
+	markerIdx := len(batch.Checksums) - 1
+	for {
+		// NB: If the incoming checksum block is empty, move to the next one.
+		if len(batch.Checksums) == 0 {
+			if !c.blockReader.Next() {
+				c.exhausted = true
+				return c.mismatches, nil
+			}
+
+			batch = c.blockReader.Current()
+			markerIdx = len(batch.Checksums) - 1
+			c.batchIdx = 0
+			continue
+		}
+
+		checksum := batch.Checksums[c.batchIdx]
+		compare := bytes.Compare(batch.EndMarker, entry.entry.ID)
+		if c.batchIdx == markerIdx {
+			// NB: this is the last element in the batch. Check ID against MARKER.
+			if entry.idChecksum == checksum {
+				if compare != 0 {
+					// Checksums match but IDs do not. Treat as invariant violation.
+					return nil, c.invariant(batch.EndMarker, checksum, entry)
+				}
+
+				// ID and checksum match. Advance the block iter and return gathered mismatches.
+				if !c.blockReader.Next() {
+					c.exhausted = true
+				} else {
+					batch = c.blockReader.Current()
+					markerIdx = len(batch.Checksums) - 1
+					c.batchIdx = 0
+				}
+
+				return c.mismatches, nil
+			}
+
+			// Checksum mismatch.
+			if compare == 0 {
+				// IDs match but checksums do not. Advance the block iter and return
+				// mismatch.
+				if !c.blockReader.Next() {
+					c.exhausted = true
+				} else {
+					batch = c.blockReader.Current()
+					markerIdx = len(batch.Checksums) - 1
+					c.batchIdx = 0
+				}
+
+				return c.entryMismatches(entry), nil
+			} else if compare > 0 {
+				// This is a mismatch on primary that appears before the
+				// marker element. Return mismatch but do not advance iter.
+				return c.entryMismatches(entry), nil
+			}
+
+			// The current batch here is exceeded. Emit the current batch marker as
+			// a mismatch on primary, and advance the block iter.
+			c.indexMismatches(checksum)
+			if !c.blockReader.Next() {
+				// If no further values, add the current entry as a mismatch and return.
+				c.exhausted = true
+				return c.entryMismatches(entry), nil
+			}
+
+			batch = c.blockReader.Current()
+			markerIdx = len(batch.Checksums) - 1
+			c.batchIdx = 0
+			continue
+		}
```

Comment:
A comment indicating this is the end of `if c.batchIdx == markerIdx` would be nice here. There's simply too much scrolling to match all those ifs.